### ISSUE TITLE
Error: Could not set 'present' on ensure: undefined method `[]' for nil:NilClass

### ISSUE THREAD

**Issue by Skarzee** [logicmonitor/logicmonitor-puppet-v4#13]
Hey guys,

We're running into a few issues here at BookingGo/Rentalcars recently regarding the module and I'm a bit at odds as to what it is. The requests seems to be evaluating incorrectly, particularly on checking if data exists.

# Environment:

I've tried a few things already to see if this is on our side and a machine error, this includes multiple versions of Ruby. I essentially have 2 collectors at present and they are separated like so:

## Environment 1 (Alfa):
```
[Thu Mar 01 15:11:46] [root@lm-collector-01.dqs agent]# cat /etc/centos-release
CentOS Linux release 7.3.1611 (Core)

[Thu Mar 01 15:09:49] [root@lm-collector-01.dqs ~]# ruby -v
ruby 1.9.3p551 (2014-11-13 revision 48407) [x86_64-linux]

Installed Path:
/usr/local/logicmonitor/agent/

Puppet version: 
3.7.3

PuppetDB Enabled/Installed:
YES

```

## Environment 2 (Bravo):
```
[Thu Mar 01 15:05:00] [root@lm-collector-02.ams4 agent]# cat /etc/centos-release
CentOS Linux release 7.4.1708 (Core)

[Thu Mar 01 14:59:07] [root@lm-collector-02.ams4 agent]# ruby -v
ruby 2.0.0p648 (2015-12-16) [x86_64-linux]

Installed Path:
/usr/local/logicmonitor/agent/

Puppet version: 
3.7.3

PuppetDB Enabled/Installed:
NO
```

### Error: Alfa

Environment 1 (Alfa) has successfully registered itself in the past as it has PuppetDB enabled, however we are currently not running PuppetDB in production. The second node has never therefore been registered, but was running previously without errors in the past. This environment has PuppetDB, therefore we are able to register nodes, this is me trying to register the device:

```
Debug: Device[watdev-01.dqs.traveljigsaw.com](provider=device): Creating device: "watdev-01.dqs.traveljigsaw.com"
Error: Could not set 'present' on ensure: undefined method `[]' for nil:NilClass
Error: Could not set 'present' on ensure: undefined method `[]' for nil:NilClass
Wrapped exception:
undefined method `[]' for nil:NilClass
Error: /Stage[main]/Logicmonitor::Master/Device[watdev-01.dqs.traveljigsaw.com]/ensure: change from absent to present failed: Could not set 'present' on ensure: undefined method `[]' for nil:NilClass
```

### Error: Bravo
I put an extra debug in to eval the FQDN and that returns fine

```
Debug: Collector[lm-collector-02.ams4.traveljigsaw.com](provider=collector): Checking if Collector exists with description "lm-collector-02.ams4.traveljigsaw.com"
Debug: Collector[lm-collector-02.ams4.traveljigsaw.com](provider=collector): Checking contents of description... "lm-collector-02.ams4.traveljigsaw.com"
Error: /Stage[main]/Logicmonitor::Install/Collector[lm-collector-02.ams4.traveljigsaw.com]: Could not evaluate: no implicit conversion of nil into String
Notice: /Stage[main]/Logicmonitor::Install/Collector_installer[lm-collector-02.ams4.traveljigsaw.com]: Dependency Collector[lm-collector-02.ams4.traveljigsaw.com] has failures: true
Warning: /Stage[main]/Logicmonitor::Install/Collector_installer[lm-collector-02.ams4.traveljigsaw.com]: Skipping because of failed dependencies
Notice: /Stage[main]/Logicmonitor::Service/Service[logicmonitor-agent]: Dependency Collector[lm-collector-02.ams4.traveljigsaw.com] has failures: true
Warning: /Stage[main]/Logicmonitor::Service/Service[logicmonitor-agent]: Skipping because of failed dependencies
Notice: /Stage[main]/Logicmonitor::Service/Service[logicmonitor-watchdog]: Dependency Collector[lm-collector-02.ams4.traveljigsaw.com] has failures: true
Warning: /Stage[main]/Logicmonitor::Service/Service[logicmonitor-watchdog]: Skipping because of failed dependencies.
```

Do you have an input or suggestions for this? As something seems inherently wrong. I can't see much within the provider, but it seems that the POST is having difficulties actually being transmitted.  These runs are failing fairly recently, but I cannot manage to marry up any of the recent commits.

Thanks!

**Comment by samjdacanay** [comment]
Hi @Skarzee ,

With respect to error alpha: This could be an issue related to how the device is defined on the manifest. Furthermore, for those logs, would it be possible to see any additional stacktrace that is printed?

For error bravo, we expect this, because puppet DB is a dependency of our integration module.

Thanks for reporting this issue.

- Sam

**Comment by Skarzee** [comment]
Hey guys, I believe this can be closed now. I managed to get some time in today to work through these issues, and came up with the following realisation (both of the below points are with Alfa):

+ We have weird scoping set within our modules via an ENC. This means that we need to scope explicitly, as scoping is set, by default, to our company specific module path. This means that the above was unable to call the device class correctly, which makes sense now. We had to change this to `::logicmonitor::device`. Afterwards, the class =! to nil.
+ Additionally, something else that I've learnt is that the exporting of resources to Puppet DB, is that they seem to be quiet when exporting to PuppetDB. So if you're looking to debug ( `puppet agent -td`) you're not going to see anything. Best way to do this debugging is to query the PuppetDB API directly: ``curl -X GET http://puppetdb:8080/v3/resources  --data-urlencode 'query=["=","exported", true]'``
+ Thirdly, the work that is done, as discussed via email, is done only when the collector runs Puppet. I think this would be good to highlight. The Puppet run on a node (device) would simply store it to PuppetDB, and the collector would then harvest the devices and do the API communication to limit API calls as this is seemingly done through a single connection.

Thanks for you attention on this though!
